src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/DiagnosticSourceManager.cs

```diff
@@ -0,0 +1,91 @@
+        private ImmutableDictionary<string, ImmutableArray<IDiagnosticSourceProvider>>? _workspaceSources;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public DiagnosticSourceManager([ImportMany] Lazy<IEnumerable<IDiagnosticSourceProvider>> sources)
+        {
+            _sources = sources;
+        }
+
+        /// <inheritdoc />
+        public IEnumerable<string> GetSourceNames(bool isDocument)
+        {
+            EnsureInitialized();
+            return (isDocument ? _documentSources : _workspaceSources)!.Keys;
+        }
+
+        /// <inheritdoc />
+        public async ValueTask<ImmutableArray<IDiagnosticSource>> CreateDiagnosticSourcesAsync(RequestContext context, string sourceName, bool isDocument, CancellationToken cancellationToken)
+        {
+            EnsureInitialized();
+            var providersDictionary = isDocument ? _documentSources : _workspaceSources;
+            if (!providersDictionary!.TryGetValue(sourceName, out var providers))
+            {
+                return [];
+            }
+
+            using var _ = ArrayBuilder<IDiagnosticSource>.GetInstance(out var builder);
+            foreach (var provider in providers)
+            {
+                var sources = await provider.CreateDiagnosticSourcesAsync(context, sourceName, cancellationToken).ConfigureAwait(false);
+                builder.AddRange(sources);
+            }
+
+            return builder.ToImmutableAndClear();
+        }
+
+        private void EnsureInitialized()
+        {
+            if (_documentSources == null || _workspaceSources == null)
+            {
+                Dictionary<string, IReadOnlyList<IDiagnosticSourceProvider>> documentSources = new();
+                Dictionary<string, IReadOnlyList<IDiagnosticSourceProvider>> workspaceSources = new();
+                foreach (var source in _sources.Value)
+                {
+                    var attribute = source.GetType().GetCustomAttributes<ExportDiagnosticSourceProviderAttribute>(inherit: false).FirstOrDefault();
+                    if (attribute != null)
+                    {
+                        var scopedSources = source.IsDocument ? documentSources : workspaceSources;
+                        foreach (var sourceName in source.SourceNames)
+                        {
+                            if (!scopedSources.TryGetValue(sourceName, out var sources))
+                            {
+                                sources = new List<IDiagnosticSourceProvider>();
+                                scopedSources[sourceName] = sources;
+                            }
+                            ((List<IDiagnosticSourceProvider>)sources).Add(source);
+                        }
+                    }
+                }
+
+                var immutableSources = documentSources.ToImmutableDictionary(entry => entry.Key, entry => entry.Value.ToImmutableArray());
```

Comment:
could these just be a lazy instead?

Reply:
Turned out there is no point to make anything lazy here as we hit this code path during init sequence (to inform host about available diag sources)